### PR TITLE
Creates AI Widget sub module

### LINKED ISSUE #1: Create AI Widget module
As a web manager

I need to place Netcall code at specific points in a page, and in specific pages

So that the correct Netcall assistant will run

Functionality needs to be part of the new localgov_netcall module that Adnan is creating, so Netcall councils only have to install and update one thing

### PR DESCRIPTION
**Pull request by markconroy**
Closes #1 

---

1. Settings page at /admin/config/services/netcall/ai-widget
2. Field for "workspace id"
3. Field for "partition id"
4. Field for "partition overrides

## Settings page
Go to the settings page to fill out your settings. Remember, these will become part of your site's config, so if you change these you need a developer to export your config so they are not wiped the next time a deployment happens.

## Field for workspace id
Put the workspace id in here. It becomes site config, so a developer will be needed to capture that.

## Field for partition id
Put the partition id in here. It becomes site config, so a developer will be needed to capture that.

## Good to go
This will be enough for most sites, if they only have one ai widget. 

## Field for partition overrides
If you want to override what widget is show on what page you can add a list of different partition ids and paths to this field, in the format of `partition_id|path` (such as `par123abc|news` for `par123abc` to appear on the news page, but the default ai widget to appear everywhere else).

If you want a different widget to appear on a section of pages, such as all the items inside `/news` put a `*` at the end of your path like so: `par123abc|news/*`

Screenshot 2024-08-05 at 13 10 11

---
Thanks to [Big Blue Door](https://www.bigbluedoor.net/) for sponsoring my time to work on this.

### CODE REVIEW

**Comment by markconroy** [comment]
@willguv I think this module is ready for testing/feedback now.

File: modules/localgov_netcall_ai_widget/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+## INTRODUCTION
+
+The LocalGov Netcall AI Widget module is a DESCRIBE_THE_MODULE_HERE.
```
Comment (Adnan-cds) [inline modules/localgov_netcall_ai_widget/README.md]: Err.. quite a few placeholders in this file.  If you can please look into them.

File: modules/localgov_netcall_ai_widget/config/schema/localgov_netcall_ai_widget.schema.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Schema for the configuration files of the Content Access module.
+localgov_netcall_ai_widget.settings:
+  type: config_object
+  label: 'Netcall AI Widget settings'
+  mapping:
+    roles:
+      type: checkboxes
```
Comment (Adnan-cds) [inline modules/localgov_netcall_ai_widget/config/schema/localgov_netcall_ai_widget.schema.yml]: As far as I remember, the `type` here refers to a data type.  [Potential values](https://www.drupal.org/docs/drupal-apis/configuration-api/configuration-schemametadata#types) include boolean, integer, etc.  Please try the [config_inspector](https://www.drupal.org/project/config_inspector) module to validate this.  If config_inspector spots an error, try the following snippet which is from the [views](https://git.drupalcode.org/project/drupal/-/blob/11.x/core/modules/views/config/schema/views.data_types.schema.yml?ref_type=heads#L782) module:
```
roles:
  type: sequence
  label: 'User roles'
  sequence:
    type: string
    label: 'Role'
```

File: modules/localgov_netcall_ai_widget/js/ai-widget-init.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                (r.src = e.url),
+                (n = s.parentNode) == null || n.insertBefore(r, s);
+            }
+          } catch {}
+        }
+        const o = new XMLHttpRequest();
+        o.addEventListener('load', a),
+          o.open(
+            'GET',
+            `https://webassist.netcall-apollo-dev.co.uk/api/v1/loader/workspaces/${workspaceId}/definitions/${partitionId}`,
```
Comment (Adnan-cds) [inline modules/localgov_netcall_ai_widget/js/ai-widget-init.js:99]: Please confirm this URL.

File: modules/localgov_netcall_ai_widget/js/ai-widget-init.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      const currentPath = window.location.pathname;
+      const partitionOverrides = [];
+      const partitionOverridesFromConfig =
+        drupalSettings.localgovNetcall.aiWigdetPartitionOverrides;
+
+      // Check if there's any overrides in the config.
+      // If so, add each to the partitionOverrides array.
+      if (partitionOverridesFromConfig) {
+        partitionOverridesFromConfig
+          .split('\r\n')
```
Comment (Adnan-cds) [inline modules/localgov_netcall_ai_widget/js/ai-widget-init.js:22]: I hope `\r\n` works for everybody :)

**Review by Adnan-cds — COMMENTED** [review]
Hi Mark,
I am yet to test this.  I have highlighter some minor issues in the code.  I will try to provide more feedback once I figure out how to test this.

File: modules/localgov_netcall_ai_widget/js/ai-widget-init.js (hunk 3 above)
Comment (markconroy) [reply modules/localgov_netcall_ai_widget/js/ai-widget-init.js:99]: This is the URL from the code snippet I was given.

File: modules/localgov_netcall_ai_widget/README.md (hunk 1 above)
Comment (markconroy) [reply modules/localgov_netcall_ai_widget/README.md]: Ooops, thanks Adnan.

File: modules/localgov_netcall_ai_widget/js/ai-widget-init.js (hunk 4 above)
Comment (markconroy) [reply modules/localgov_netcall_ai_widget/js/ai-widget-init.js:22]: If we get any complaints, we can figure it out then.

File: modules/localgov_netcall_ai_widget/config/schema/localgov_netcall_ai_widget.schema.yml (hunk 2 above)
Comment (markconroy) [reply modules/localgov_netcall_ai_widget/config/schema/localgov_netcall_ai_widget.schema.yml]: Thanks Adnan, a great spot from you. This should be fixed via a00b18a

**Comment by Adnan-cds** [comment]
Hi Mark,
Code changes look good.  I am just trying to test this new module.  I will get back to you by the end of the day.

**Review by Adnan-cds — APPROVED** [review]
Thank you.  I can see the configuration values appearing in the browser's Javascript console under `drupalSettings.localgovNetcall`.

**Comment by finnlewis** [comment]
@Adnan-cds  is happy, but not been able to test fully yet.

Happy to merge, and follow up if needed.

**Pull request merged**